src/ucp/core/ucp_am.c

```diff
@@ -1021,6 +1021,14 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_am_recv_data_nbx,
     ucs_status_t status;
     size_t recv_length;
 
+    if (ENABLE_PARAMS_CHECK &&
+        !ucs_unlikely((desc->flags & UCP_RECV_DESC_FLAG_UCT_DESC) ||
+                      (desc->flags & UCP_RECV_DESC_FLAG_RNDV))) {
+        ucs_error("attempt to data receiving for the wrong descriptor");
+        ret = UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM);
+        goto out;
```

Comment:
please either return here immdeiately, or move this check after `UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);`, 
because going to `out` section assumes releasing of worker lock.